### PR TITLE
Algorithm: betweenness centrality WIP

### PR DESCRIPTION
**Pull request by mknblch**
**WIP** please do not merge yet.

I found a promising implementation of BC in [this paper](http://www.algo.uni-konstanz.de/publications/b-fabc-01.pdf) and wanted to give it a try. 

To add an appropriate testcase i had to merge my other WIP branch with the TestGraphBuilders. We can split those things into 2 PRs if you want @jexp.

**Note**: I added default ctor to `org.neo4j.graphalgo.core.sources.LazyIdMapper` which makes the computation of `nodeCount()` indeed lazy. Using the nodeCount constructor still behaves as usual.

**Todo**: 
- The paper also describes how to implement other centrality measurements using this method - implement!
- check if we can use node partitioning for parallelization

### CODE REVIEW

File: algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ * @author mknblch
+ */
+public class BetweennessCentrality {
+
+    private final Graph graph;
+
+    private final double[] centrality;
+    private final double[] delta;
+    private final int[] sigma;
+    private final int[] d;
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java]: can you use a better name for `d`?

File: algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        sigma[startNode] = 1;
+        d[startNode] = 0;
+        queue.addLast(startNode);
+        while (!queue.isEmpty()) {
+            int node = queue.removeLast();
+            stack.push(node);
+            graph.forEachRelationship(node, Direction.OUTGOING, (source, target, relationId) -> {
+                if (d[target] < 0) {
+                    queue.addLast(target);
+                    d[target] = d[node] + 1;
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java]: we could compute d[node]+1 once before the loop and reuse it.

File: algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                }
+                return true;
+            });
+        }
+        while (!stack.isEmpty()) {
+            final int node = stack.pop();
+            if (null == paths[node]) {
+                continue;
+            }
+            paths[node].forEch(v -> {
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java]: forEach typo?

Comment (mknblch) [inline algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java]: ouh its late ;)

File: algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    /**
+     * append nodeId to path
+     *
+     * @param path the selected path
+     * @param nodeId the node id
+     */
+    private void append(int path, int nodeId) {
+        if (null == paths[path]) {
+            paths[path] = new Path();
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java]: can we / do we know a good size to initialize the path with?

File: core/src/main/java/org/neo4j/graphalgo/core/utils/container/Path.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * @author mknblch
+ */
+public class Path {
+
+    private int[] nodes;
+    private int offset = 0;
+
+    public Path() {
+        this(1); // TODO test other initial sizes
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/container/Path.java]: probably 10 or similar common path lengths?

File: algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java (hunk 4 above)
Comment (mknblch) [inline algo/src/main/java/org/neo4j/graphalgo/impl/BetweennessCentrality.java]: This is just my first poc. I'd like to implement paths on a preallocated array if possible. but lets have a talk with @knutwalker before

File: core/src/main/java/org/neo4j/graphalgo/core/utils/container/Path.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public Path(int initialSize) {
+        nodes = new int[initialSize];
+    }
+
+    public void append(int nodeId) {
+        if (offset >= nodes.length) {
+            nodes = ArrayUtil.grow(nodes, offset);
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/container/Path.java]: optimization for the future, have ArrayUtil pool reused arrays?

File: tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        return me();
+    }
+
+    public ME setRelationship(String relationship) {
+        this.relationship = relationship;
+        return me();
+    }
+
+    public Relationship createRelationship(Node p, Node q) {
+        return p.createRelationshipTo(q, RelationshipType.withName(relationship));
```
Comment (jexp) [inline tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java]: store rel-type in instance variable

File: tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    public Node createNode() {
+        Node node = api.createNode();
+        if (null != label) {
+            node.addLabel(Label.label(label));
```
Comment (jexp) [inline tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java]: store label in instance variable

File: tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        tx = api.beginTx();
+    }
+
+    protected void closeTx() {
+        tx.success();
+        tx.close();
+        tx = null;
+    }
+
+    protected abstract ME me();
```
Comment (jexp) [inline tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java]: perhaps we can just create a field ME and store `this.me = (ME)this` in the constructor?

**Comment by jexp** [comment]
Looks good thanks a lot. Let me know when you think it's ready for a merge.

The first part of the computation looks like regular SSSP (perhaps we can reuse it there) but the sigma part is different.

File: tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java (hunk 9 above)
Comment (mknblch) [inline tests/src/main/java/org/neo4j/graphalgo/core/graphbuilder/GraphBuilder.java]: perhaps i got u wrong but me() must always return the child class instance or an invocation of an method of the abstract builder would not return the concrete type but the abstract builder.

**Comment by jexp** [comment]
I think it makes sense to merge the test graph builders first separately.
And perhaps the LazyIdMapper too.

So that those could be used elsewhere while you work on this WIP.

**Comment by mknblch** [comment]
**TODO**: looks like that BC always produces integer values. check if we can switch to int[] instead of double[].

**Comment by mknblch** [comment]
i separated the code into 2 different PRs therefore im going to close this one

**Pull request closed without merging**